kafka/src/main/java/zipkin2/reporter/kafka/KafkaReporter.java

```diff
@@ -0,0 +1,327 @@
+     *
+     * <p>Note: If ultimately sending to Zipkin, version 2.8+ is required to process protobuf.
+     */
+    public KafkaReporter.Builder encoding(Encoding encoding) {
+      if (encoding == null) throw new NullPointerException("encoding == null");
+      this.encoding = encoding;
+      return this;
+    }
+
+    public KafkaReporter build() {
+      return new KafkaReporter(this);
+    }
+  }
+
+  final Properties properties;
+  final String topic;
+  final Encoding encoding;
+  final SpanBytesEncoder encoder;
+  final int messageMaxBytes;
+  final ReporterMetrics metrics;
+
+  /** Tracks if we should log the first instance of an exception in flush(). */
+  private boolean shouldWarnException = true;
+
+  public static KafkaReporter create(String bootstrapServers) {
+    return newBuilder().bootstrapServers(bootstrapServers).build();
+  }
+
+  public static KafkaReporter.Builder newBuilder() {
+    // Settings below correspond to "Producer Configs"
+    // http://kafka.apache.org/0102/documentation.html#producerconfigs
+    Properties properties = new Properties();
+    properties.put(ProducerConfig.KEY_SERIALIZER_CLASS_CONFIG, ByteArraySerializer.class.getName());
+    properties.put(ProducerConfig.VALUE_SERIALIZER_CLASS_CONFIG,
+      ByteArraySerializer.class.getName());
+    // 1MB, aligned with default kafka max.message.bytes config.
+    properties.put(ProducerConfig.MAX_REQUEST_SIZE_CONFIG, 500000);
+    properties.put(ProducerConfig.ACKS_CONFIG, "0");
+    return new KafkaReporter.Builder(properties);
+  }
+
+  KafkaReporter(KafkaReporter.Builder builder) {
+    properties = new Properties();
+    properties.putAll(builder.properties);
+    topic = builder.topic;
+    encoding = builder.encoding;
+    switch (encoding) {
+      case JSON:
+        encoder = SpanBytesEncoder.JSON_V2;
+        break;
+      case PROTO3:
+        encoder = SpanBytesEncoder.PROTO3;
+        break;
+      case THRIFT:
+        encoder = SpanBytesEncoder.THRIFT;
+        break;
+      default:
+        throw new UnsupportedOperationException(encoding.name());
+    }
+    messageMaxBytes = builder.messageMaxBytes;
+    metrics = builder.metrics;
+  }
+
+  public KafkaReporter.Builder toBuilder() {
+    return new KafkaReporter.Builder(this);
+  }
+
+  /**
+   * get and close are typically called from different threads
+   */
+  volatile KafkaProducer<byte[], byte[]> producer;
+  volatile boolean closeCalled;
+  volatile AdminClient adminClient;
+
+  @Override public void report(Span span) {
+    if (closeCalled) throw new IllegalStateException("closed");
+    metrics.incrementSpans(1);
+    int spanSizeInBytes = encoder.sizeInBytes(span);
+    metrics.incrementSpanBytes(spanSizeInBytes);
+    ProducerRecord<byte[], byte[]> record =
+      new ProducerRecord<>(
+        topic,
+        span.traceId().getBytes(),
+        encoder.encodeList(Collections.singletonList(span)));
+    getProducer().send(record, (metadata, exception) -> {
```

Comment:
ps while not opposed to this, I also think a kafka "FinishedSpanHandler" in brave could be more effective. you could bundle messages by local root. since we don't have a field in span for this, it could only be implemented in brave.

Reply:
>  I also think a kafka "FinishedSpanHandler" in brave could be more effective. you could bundle messages by local root. 

This is something I miss from `AsyncReporter` where I can group per `traceId`, and send them together.

local root would solve the same but from brave, right?

> kafka "FinishedSpanHandler" in brave

Haven't worked with this API yet tho, but a `kafka handler` would replace `reporter` under site-specific config plus grouping spans per local root, right? 

I'm just a bit confused on the overlapping between handlers and reporter, and if it's ok to solve this on both sides or pick one.